README.md
MariaDB-Manager-Monitor
=======================

MariaDB-Manager-Monitor is the monitoring component of the MariaDB-Manager project. It provides a confirgurable monitoring service in Java that interacts with the MariaDB-Manager-API to provide monitoring data on a set of servers within a MariaDB Galera Cluster.

The monitor requires another component of the MariaDB-Manager family in order to be built, libMariaDB-Manager-java. This project provides library functions that are shared between the Java components of MariaDB-Manager.

Other components within the MariaDB-Manager family are

| Project                   | Description                                                                |
|---------------------------|----------------------------------------------------------------------------|
| MariaDB-Manager           | The overall project that encapsualtes all the components of the MariaDB-Manager|
| MariaDB-Manager-API       | The REST API that gives access to all MariaDB-Manager monitoring and management features|
| MariaDB-Manager-GREX      | The remotely executed components of the MariaDB-Manager that are installed on every node in the clsuter|
| MariaDB-Manager-WebUI     | The web user interface for the MariaDB-Manager                             |



ReleaseNotes.txt
#  Part of MariaDB Manager package.
#
# This file is distributed as part of MariaDB Manager. It is free
# software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation,
# version 2.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright 2013-2014 (c) SkySQL Corporation Ab
#
# Author: Massimo Siani
# Date: August 2013
#




CHANGELOG

* Tue May 13 2014 Massimo Siani <massimo.siani@skysql.com>
- LSBize mariadb-manager-monitor

* Tue Apr 29 2014 Massimo Siani <massimo.siani@skysql.com>
- Fix .spec

* Mon Apr 28 2014 Massimo Siani <massimo.siani@skysql.com>
- Do not check for /etc/mariadbmanager/manager.ini

* Fri Apr 25 2014 Massimo Siani <massimo.siani@skysql.com>
- Small script fix.
- Remove a node object from the Global Object when the API returns it again.
- Debian: do not create the configuration file, nor its folder.

* Thu Apr 24 2014 Massimo Siani <massimo.siani@skysql.com>
- Fix spelling.

* Wed Apr 23 2014 Massimo Siani <massimo.siani@skysql.com>
- Created release-1.0.2

* Thu Apr 17 2014 Massimo Siani <massimo.siani@skysql.com>
- Create section [monitor] in /etc/mariadbmanager/manager.ini if does not exist.
- Update API key generation scripts.

* Wed Apr 16 2014 Massimo Siani <massimo.siani@skysql.com>
- Load all the current configuration options from /etc/mariadbmanager/manager.ini

* Tue Apr 15 2014 Massimo Siani <massimo.siani@skysql.com>
- Start implementing support for manager.ini

* Mon Apr 14 2014 Massimo Siani <massimo.siani@skysql.com>
- Add the version date, and register it in the API.
- Naming now follows standard conventions.
- Move the API client to the Library.

* Fri Apr 11 2014 Massimo Siani <massimo.siani@skysql.com>
- Fixes for Debian
- Manage Java 6
- Merge debian branch into develop

* Tue Apr 08 2014 Massimo Siani <massimo.siani@skysql.com>
- Rename mariadb-enterprise-monitor to mariadb-manager-monitor
- There is no need to update the version every 30 seconds, set to 5 minutes

* Thu Mar 27 2014 Massimo Siani <massimo.siani@skysql.com>
- Add database information with the system state
- Update libMariaDB-Manager-java dependency

* Wed Mar 26 2014 Massimo Siani <massimo.siani@skysql.com>
- Register the release number in the API, register the Java Library

* Wed Mar 12 2014 Massimo Siani	<massimo.siani@skysql.com>
- Change MariaDB Enterprise to MariaDB Manager

* Thu Feb 27 2014 Massimo Siani <massimo.siani@skysql.com>
- Merge develop and release-1.0.1

* Fri Feb 21 2014 Massimo Siani	<massimo.siani@skysql.com>
- Fix Copyright and files headers

* Thu Feb 20 2014 Massimo Siani	<massimo.siani@skysql.com>
- Bump to version 1.7-121

* Tue Feb 18 2014 Massimo Siani	<massimo.siani@skysql.com>
- Fix pom.xml
- Merge Timofey's changes

* Thu Feb 13 2014 Massimo Siani	<massimo.siani@skysql.com>
- Deleted duplicate pom.xml

* Wed Feb 12 2014 Massimo Siani <massimo.siani@skysql.com>
- Fix API call for bulk data

* Fri Dec 13 2013 Massimo Siani <massimo.siani@skysql.com>
- Do not start the Monitor after installation

* Wed Dec 11 2013 Massimo Siani <massimo.siani@skysql.com>
- Add key generation upon install

* Fri Nov 29 2013 Massimo Siani <massimo.siani@skysql.com>
- Fix headers date

* Thu Nov 28 2013 Massimo Siani <massimo.siani@skysql.com>
- Change API call monitorclass/galera/key to monitorclass/galera

* Thu Nov 21 2013 Massimo Siani <massimo.siani@skysql.com>
- Fix headers
- Fix bug on Monitor version

* Tue Nov 19 2013 Massimo Siani <massimo.siani@skysql.com>
- Register Monitor version every 30 seconds
- Added machine-down state for Ping monitor
- Fixed locale on non-English machines

* Fri Nov 15 2013 Massimo Siani <massimo.siani@skysql.com>
- Added version registration in the API
- Include a modified mariadb java driver to avoid select commands on connect

* Tue Nov 12 2013 Massimo Siani <massimo.siani@skysql.com>
- Use show status instead of select

* Thu Nov 7 2013 Massimo Siani <massimo.siani@skysql.com>
- Split library

* Fri Nov 1 2013 Massimo Siani <massimo.siani@skysql.com>
- Fix compatibility with API R185

* Mon Oct 28 2013 Massimo Siani <massimo.siani@skysql.com>
- Limit the information request in the JSON

* Tue Oct 23 2013 Massimo Siani <massimo.siani@skysql.com>
- Fix bug after node removal

* Mon Oct 21 2013 Massimo Siani <massimo.siani@skysql.com>
- Double check private IP of the node before connecting.
- Throw a human readable message from JS monitor if the node is down.

* Fri Oct 18 2013 Massimo Siani <massimo.siani@skysql.com>
- Check for both hostname and IP in node cluster membership

* Thu Oct 17 2013 Massimo Siani <massimo.siani@skysql.com>
- Add support for delta values in JS monitors
- Select the log verbosity by setting a variable in the init script
- Use IP if hostname is not set

* Wed Oct 16 2013 Massimo Siani <massimo.siani@skysql.com>
- Add monitor cache
- Improve configuration loading

* Thu Oct 10 2013 Massimo Siani <massimo.siani@skysql.com>
- Add init script

* Mon Oct 7 2013 Massimo Siani <massimo.siani@skysql.com>
- When a node is deleted, also remove its cached data

* Fri Oct 4 2013 Massimo Siani <massimo.siani@skysql.com>
- Fix a bug when retrieve node credentials
- Fix a bug when caching node data

* Tue Oct 1 2013 Massimo Siani <massimo.siani@skysql.com>
- Bug fix
- Fix .spec file

* Mon Sep 30 2013 Massimo Siani <massimo.siani@skysql.com>
- Add Galera nodes membership.
- Switch to syslog.
- Add provisioned node API support.
- Refresh based on If-Modified-Since.

* Tue Sep 10 2013 Massimo Siani <massimo.siani@skysql.com>

src/com/skysql/monitor/monAPI.java
/*
 * This file is distributed as part of the SkySQL Cloud Data Suite.  It is free
 * software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation,
 * version 2.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Copyright SkySQL Ab
 */

package com.skysql.monitor;


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Properties;
import java.util.Date;



import java.net.HttpURLConnection;
import java.net.URL;

/**
 * The interface to the SkySQL Manager API, a REST interface to the monitor database.
 * This is the API that should be used to store all monitor observations and to retrieve 
 * system configuration information rather than access the SQLite monitoring database directly.
 * 
 * @author Mark Riddoch
 *
 */
public class monAPI {
	/**
	 * The host on which the API is running
	 */
	String		m_apiHost;
	/**
	 * The API Key to use to access the API
	 */
	String		m_apiKey;
	/**
	 * The API key ID as assigned by the API provider
	 */
	String		m_apiKeyID;
	
	/**
	 * Construct the monAPI instance. This consists of obtaining the information required
	 * to contact the API. This information is available via Java System Properties.
	 */
	public monAPI()
	{
		Properties props = System.getProperties();
		m_apiHost = props.getProperty("SKYSQL_API_HOST", "localhost");
		m_apiKey = props.getProperty("SKYSQL_API_KEY", "1f8d9e040e65d7b105538b1ed0231770");
		m_apiKeyID = props.getProperty("SKYSQL_API_KEYID", "1");
	}
	
	/**
	 * Populate a monitor value for a given node.
	 * 
	 * @param systemID	The ID of the System
	 * @param nodeID	The ID of the node within the system that the data refers to
	 * @param monitorID	The ID of the monitor itself
	 * @param value		The observed value
	 * @return True if the update was performed to the API
	 */
	public boolean MonitorValue(int systemID, int nodeID, int monitorID, String value)
	{
		
		return restPost("system/" + systemID + "/node/" + nodeID + "/monitor/" + monitorID + "/data", "value", value);
	}
	
	/**
	 * Populate a monitor value for the system
	 * 
	 * @param systemID	The ID of the System
	 * @param monitorID	The ID of the monitor itself
	 * @param value		The observed value
	 * @return True if the update was performed to the API
	 */
	public boolean MonitorValue(int systemID, int monitorID, String value)
	{
		return restPost("system/" + systemID + "/monitor/" + monitorID + "/data", "value", value);
	}


	/**
	 * Send a POST request to the API
	 * 
	 * @param restRequest	The URL, excluding the fixed stem
	 * @param pName			The parameter name for the post request
	 * @param pValue		The parameter value for the port request
	 */
	private boolean restPost(String restRequest, String pName, String pValue)
	{
		String result = null;
		
		String value = pName + "=" + pValue;
		
		
		try {
	 
			String reqString = "http://" + m_apiHost + "/consoleAPI/api/" + restRequest;
			URL postURL = new URL(reqString);
			HttpURLConnection apiConn = (HttpURLConnection)postURL.openConnection();
			apiConn.setRequestMethod("POST");
			SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z");
		    String rfcdate = sdf.format(new Date());
			
			String fullkey = restRequest + m_apiKey + rfcdate;
			MessageDigest md = MessageDigest.getInstance("MD5");

	        md.update(fullkey.getBytes());
	        byte byteData[] = md.digest();
	        //convert the byte to hex format method 1
	        StringBuffer sb = new StringBuffer();
	        for (int i = 0; i < byteData.length; i++) {
				sb.append(Integer.toString((byteData[i] & 0xff) + 0x100, 16).substring(1));
	        }

			apiConn.addRequestProperty("Date", rfcdate);
			apiConn.setRequestProperty("Authorization", "api-auth-" + m_apiKeyID + "-" + sb.toString());
			apiConn.setRequestProperty("Accept", "application/json");
			apiConn.setDoOutput(true);
			apiConn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
			apiConn.setRequestProperty("charset", "utf-8");
			apiConn.setRequestProperty("Content-Length", "" + Integer.toString(value.getBytes().length));
			apiConn.setUseCaches(false);
			apiConn.setRequestMethod("POST");
			OutputStreamWriter out = new OutputStreamWriter(apiConn.getOutputStream());
			out.write(value);
			out.close();
			
			BufferedReader in = new BufferedReader(new InputStreamReader(apiConn.getInputStream()));
			result = in.readLine();
			in.close();
			
			if (apiConn.getResponseCode() != 200) {
				throw new RuntimeException("Failed : HTTP error : "
					+ apiConn.getResponseMessage() + ": returned data: " + result);
			}
		  } catch (IOException e) {
					e.printStackTrace();
					return false;
		  } catch (Exception e) {
			  		e.printStackTrace();
			  		return false;
		  }
		return true;
	}
	
	/**
	 * Send a POST request to the API
	 * 
	 * @param restRequest	The URL, excluding the fixed stem
	 * @param pName			The parameter names for the post request
	 * @param pValue		The parameter values for the port request
	 */
	private boolean restPost(String restRequest, String pName[], String pValue[])
	{
		String result = null;
		
		/*
		 * We must have the same number of names as values
		 */
		if (pName.length != pValue.length)
		{
			return false;
		}
		
		
		try {
	 
			/*
			 * Discover the content length by finding the sum of all the pName and pValue
			 * lengths. Also allow 2 characters per value for the "=" and newline.
			 */
			int contentLength = 0;
			for (int i = 0; i < pName.length; i++)
			{
				contentLength += pName[i].length() + pValue[i].length() + 2;
			}
			String reqString = "http://" + m_apiHost + "/consoleAPI/api/" + restRequest;
			URL postURL = new URL(reqString);
			HttpURLConnection apiConn = (HttpURLConnection)postURL.openConnection();
			apiConn.setRequestMethod("POST");
			SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z");
		    String rfcdate = sdf.format(new Date());
			
			String fullkey = restRequest + m_apiKey + rfcdate;
			MessageDigest md = MessageDigest.getInstance("MD5");

	        md.update(fullkey.getBytes());
	        byte byteData[] = md.digest();
	        //convert the byte to hex format method 1
	        StringBuffer sb = new StringBuffer();
	        for (int i = 0; i < byteData.length; i++) {
				sb.append(Integer.toString((byteData[i] & 0xff) + 0x100, 16).substring(1));
	        }

			apiConn.addRequestProperty("Date", rfcdate);
			apiConn.setRequestProperty("Authorization", "api-auth-" + m_apiKeyID + "-" + sb.toString());
			apiConn.setRequestProperty("Accept", "application/json");
			apiConn.setDoOutput(true);
			apiConn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
			apiConn.setRequestProperty("charset", "utf-8");
			apiConn.setRequestProperty("Content-Length", "" + contentLength);
			apiConn.setUseCaches(false);
			apiConn.setRequestMethod("POST");
			OutputStreamWriter out = new OutputStreamWriter(apiConn.getOutputStream());
			for (int i = 0; i < pName.length; i++)
			{
				out.write(pName[i] + "=" + pValue[i] + "\n");
			}
			out.close();
			
			BufferedReader in = new BufferedReader(new InputStreamReader(apiConn.getInputStream()));
			result = in.readLine();
			in.close();
			
			if (apiConn.getResponseCode() != 200) {
				throw new RuntimeException("Failed : HTTP error : "
					+ apiConn.getResponseMessage() + ": returned data: " + result);
			}
		  } catch (IOException e) {
					e.printStackTrace();
					return false;
		  } catch (Exception e) {
			  		e.printStackTrace();
			  		return false;
		  }
		return true;
	}
}
